srml/support/src/storage/hashed/generator.rs

```diff
@@ -307,4 +347,41 @@ pub trait AppendableStorageMap<K: codec::Codec, V: codec::Codec>: StorageMap<K,
 		storage.put_raw(&k[..], &new_val);
 		Ok(())
 	}
+
+	/// Safely append the given items to the value in the storage.
+	///
+	/// `T` is required to implement `codec::EncodeAppend`.
+	fn safe_append<'a, S, I>(
+		key : &K,
+		items: &'a[I],
+		storage: &mut S,
+	) where
+		S: HashedStorage<Self::Hasher>,
+		I: codec::Encode + Clone,
+		V: codec::EncodeAppend<Item=I> + From<&'a[I]>,
+	{
+		Self::append(key, items, storage)
+			.unwrap_or_else(|_| Self::insert(key, &items.clone().into(), storage));
+	}
+}
+
+/// A storage map with a decodable length.
+pub trait DecodeLengthStorageMap<K: codec::Codec, V: codec::Codec>: StorageMap<K, V> {
+	/// Something that can provide the default value of this storage type.
+	type Default: StorageDefault<V>;
+
+	/// Read the length of the value in a fast way, without decoding the entire value.
+	///
+	/// `T` is required to implement `Codec::DecodeLength`.
+	fn decode_len<S: HashedStorage<Self::Hasher>>(key: &K, storage: &mut S) -> Result<usize, &'static str>
+		where V: codec::DecodeLength, V: Len
+	{
+		let k = Self::key_for(key);
+		if let Some(v) = storage.get_raw(&k[..]) {
+			<V as codec::DecodeLength>::len(&v).map_err(|e| e.what())
+		} else {
+			let default = Self::Default::default().ok_or("could not use default as fallback")?;
```

Comment:
I kind of feel weird returning an error in case of: `MyStorage: u32 => Option<u32>` `MyStorage::decode_len(not_existing_key) == Err("could not use default as fallback")`

At some point for me the best would be to have a different signature for storage with option and storage without:
decode_len would return `Result<Option<usize>, _>` for storage with option and `Result<usize, _>` for a storage without.

But this would need a new associated type in storagemap. I'm not sure what we want for this PR.

Reply:
I have to check via a test but I think if you have `MyStorage::decode_len` you get another compile-time error complaining that u32 does not implement `DecodeLength`.

Reply:
EDIT: `MyStorage: u32 => Option<Vec<u32>>`

Reply:
I see. Yeah, the error message could have been improved maybe. 

Basically, you are referring to these tests

https://github.com/paritytech/substrate/blob/346be4baad9286cf2c4f5ae835f6895664383737/srml/support/src/storage/storage_items.rs#L915-L925

and the flow that I see in it is:
- you want to read length
- key does not exist
- it is an option (we don't enforce `Default::default()` to be implemented for options hence I cannot assume it exists)
- the type does not have a default value given via our internal `type Default`. 
- Then you see this error.

I should note that at some point the test was:

```
assert_eq!(OptionMapVec::get(0), None);
assert_eq!(OptionMapVec::decode_len(0), Ok(0)); 
```

But I think now it is better and it is just a matter of returning a better error string. Maybe `Err("Could not read length through given default value")`?

Reply:
maybe "Value does not exist and could not use default as fallback"
but yeah ok